run_simulator.sh

```diff
@@ -0,0 +1,179 @@
+  local ret="$PIP install $REQUIREMENTS 2> /dev/null"
+  eval $ret
+}
+
+install_precommit_hooks() {
+  print_info "Installing pre-commit hooks..."
+  local cmd="cp hooks/pre-commit .git/hooks/pre-commit 2> /dev/null"
+  eval $cmd
+  local install_pre="$PRE_COMMIT install 2> /dev/null"
+  eval $install_pre
+}
+
+bootstrap() {
+  has_err=$?;create_venv
+  if [ $has_err -ne 0 ]; then
+    print_error "Unable to create env: $STYLE_BOLD$res"
+    exit
+  else
+    print_info "ENV: $VENV created."
+  fi
+  has_err=$?;install_requirements
+  if [ $has_err -ne 0 ]; then
+    clean_all
+    print_error "Error while installing dependencies for $COLOR_BOLD$PYMODULE: $has_err"
+    exit
+  else
+    print_info "Requirements installed."
+  fi
+  has_err=$?;install_precommit_hooks
+  if [ $has_err -ne 0 ]; then
+    print_warning "Error while installing pre-commit hooks $COLOR_BOLD$PYMODULE: $has_err"
+    exit
+  else
+    print_info "Bootstrap finished."
+  fi
+}
+
+clean() {
+  print_warning "Cleanning py cache..."
+  clean_cache="sudo find . -type d -name '__pycache__' -exec rm -rf {} 2> /dev/null"
+  eval $clean_cache
+  clean_test_cache="sudo find . -type d -name '*pytest_cache*' -exec rm -rf {} 2> /dev/null"
+  print_warning "Cleanning py-test cache..."
+  eval $clean_test_cache
+  clean_py_co="sudo find . -type f -name "*.py[co]" -exec rm -rf {} 2> /dev/null"
+  print_warning "Cleanning py-objects..."
+  eval $clean_py_co
+  print_info "Clean finished."
+}
+
+clean_all() {
+  print_warning "Cleanning virtual-environment..."
+  local rm_venv="rm -r $VENV 2> /dev/null"
+  eval $rm_venv
+  print_warning "Cleanning pre-commit hook..."
+  rm .git/hooks/pre-commit  2> /dev/null
+  print_info "Finished Virtual-environment cleanning."
+}
+
+run_raw() {
+  local raw_file=$1
+  local ret=""
+  ret="sudo $PYTHON $PYAPP.py -c $raw_file"
+  has_err=$?;eval $ret
+  if [ $has_err -ne 0 ]; then
+    print_error "Error while starting simulation for file: $raw_file"
+    exit
+  fi
+}
+
+run_thing() {
+  local thing_file=$1
+  local ret=""
+  print_info "Running thing for file $thing_file"
+  ret="sudo $PYTHON $PYAPP.py -t $thing_file"
+  has_err=$?;eval $ret
+  if [ $has_err -ne 0 ]; then
+    print_error "Error while starting simulation for file: $thing_file"
+    exit
+  fi
+}
+
+run_unit_tests() {
+  print_info "Running unit test for $PYMODULE"
+  ret="$PYTEST --cov=simulator $TEST_FOLDER"
+  has_err=$?;eval $ret
+  if [ $has_err -ne 0 ]; then
+    print_error "Error while executing unit tests for $PYMODULE"
+    exit
+  fi
+  print_info "Unit tests executed with success"
+}
+
+while getopts ":uabcr:t:" opt; do
+  case ${opt} in
+    b )
+      bootstrap
+      ;;
+    t )
+      thing_config="$OPTARG"
+      run_thing "$thing_config"
+      ;;
+    r )
+      raw_config="$OPTARG"
+      run_raw "$raw_config"
+      ;;
+    c )
+      clean
+      ;;
+    a )
+      clean
+      clean_all
+      ;;
+    u )
+      run_unit_tests
+      ;;
+    \? )
+      helper
```

Comment:
This approach will lead to flags being executed in sequence, which may make sense, but if the user types any nonsense (e.g.: -help) the help message will be printed several times.

My recommendation -> h) helper
                                       -> \?) printf "you don't seem to know what you are doing, please check -h for help\n"

Reply:
This won't solve the problem but will not print the help message several times.

Reply:
I haven't really messed with getopts but I think there should be a "native" solution.

P.S.: It appears that the flags are executed in the order they are passed, this may lead to unexpected behaviour.

Reply:
Here's an approach I used a WHILE ago: [gist](https://gist.github.com/lcbm/ee64772bac37eefbfc6bf2f8bfd5bf82#file-deprecated-knot-service-install-sh-L259)